### PR TITLE
[WIP] Feature/sqs

### PR DESCRIPTION
**Pull request by scottt2**
Hey @jcscottiii and @mogul!

This is the start of the work for the [18F #24 job](https://micropurchase.18f.gov/auctions/24).

Please feel free to comment on this as I progress through it. In order to support multiple services, I had to make things more generic. So, I had to touch parts of the existing RDS code (which I originally thought I could avoid).

Aspects of my code are based off [CF Platform Eng's SQS Broker](https://github.com/cf-platform-eng/sqs-broker). I almost wanted to go the route of augmenting a fork of that code to include RDS capabilities due to how they handle the AWS IAM and session info. That was **way** out of the scope of work, so I've been gentling trying to roll some of that into the existing AWS Broker.

Anyhow, please let me know any feedback. Once I have satisfied the spec, I will remove the [WIP] flag.

Thanks!
-Tyler

### CODE REVIEW

**Comment by scottt2** [comment]
Hey @jcscottiii @mogul !

So as I'm building out the database to support both instance types, I am seeing the pattern of single responsibility popping up in other repos and it has me wondering if this is the best approach to satisfy the use cases.

For instance: https://github.com/cf-platform-eng/aws-broker-boshrelease

This is code that stitches together three disparate, single concern brokers into a BOSH release. I feel like this may be an valid alternate approach to building one monolithic AWS Broker. I also see @jtwaleson 's [comment in the PR](https://github.com/cloudfoundry-community/rds-broker/pull/16#issuecomment-215698520) for the 18F AWS Broker into the RDS Broker referencing an aversion to one broker to rule them all. It's also garnered 2 upvotes.

My current path is to build N tables, one for each instance type--e.g., rds_instances, sqs_instance, etc. On create/bind/destroy we'd query to appropriate table and do the action. The logic to support both instance types is definitely complicating the code base and I foresee that complexity getting rather massive as we add more services to the broker, which is @jtwaleson 's concern.

If we follow the cf-platform-eng example above, we'd fork that guy and augment it to include the 18F version of the RDS broker and use/fork the SQS broker. There'd be a bit of clean up to do, but I think the end result would be more scaleable than a single AWS Broker.

Anyhow, I'm gonna pause here for sec and I'd love to hear your thoughts this. I will build whatever--I just want to make sure it will be useful to the teams.

Thanks!

**Comment by jcscottiii** [comment]
@scottt2,
I agree with creating independent service brokers. Ideally, I would like to create a less opinionated version of https://github.com/pivotal-cf/brokerapi (Bring your own logger, X-Request-Id headers, etc) to keep each broker that uses it very DRY which would allow us to focus on just specific broker code. The whole exercise of creating this monolithic one to begin with would help us tear it down into the independent brokers later on once we identify the commonalities to tear them apart. However, if it's getting icky now, I can see splitting it out now would be best.

Thoughts @dlapiduz @mogul?

**Comment by scottt2** [comment]
@jcscottiii The modularity definitely jives with me more, but I can see the argument for going all into one as well.

It's starting to get a bit tangled. Right now, everything is very much built for RDS (makes sense, it was the RDS Broker and is becoming the AWS Broker), so it requires building the generic infrastructure, refactoring the existing RDS code to use that, then introduce the SQS functionality on top of that.

I started by trying to refactor in place to being more generic, but I think the approach--if we want to continue on the monolith--should be to write the generic parts fresh--new controllers, new brokerDB code (instances table needs type, etc), better separation of the IAM and AWS session components (pull out of RDS code and make available as it's own package)--as a start, use what we can of the existing RDS broker in this framework (most the code), and then add in SQS functionality.

The question I had in starting down that path is "does this added complexity make things easier?" I don't think it does. In a world with discrete service brokers, a user would just have to "cf create-service-broker" two times instead of one, which is additional overhead for them, but vastly less overhead for developers.

Love this discussion. I am new to the Cloud Foundry world and am trying to soak up as much in terms of best practices that I can.

**Comment by scottt2** [comment]
And +1 for the https://github.com/pivotal-cf/brokerapi template. I agree--we could use that to build out whatever broker imaginable and standardize on a general config that could be used in any 18F broker based off that.

**Comment by dlapiduz** [comment]
We were having this same discussion this other day and to be honest I don't have a _great_ answer but I have _opinions_.

Having brokers dedicated for each service means that there is less interrelation between them and it is true the cloud foundry makes it very easy to install and work with them.

That said, the code required to do the RDS broker is very similar to the one required to do the SQS broker or S3 or any other AWS service. 

I don't think it is a good idea to have multiple brokers treating IAM roles in different ways for example. Or if there is an update to the API I don't want to have to patch 10 different projects.

We could extract all of that out into dependencies but at the end of the day we would still have to manage multiple projects.

I'd say lets try it out with SQS and see what happens. If it really tangles everything up and adds a bunch of stuff that makes things worse then lets split it.

At the end of the day I'd love to end up with a broker that allows people to add AWS services with just a few lines of code and that they can rest assured that 90% of the solution is already there and working.

**Comment by scottt2** [comment]
Yeah. I don't have what I feel is *the* answer. I definitely see your point about overlap of certain responsibilities that are common across all AWS services.

I can take a crack at what I specced out in terms of the monolith. I'd love feedback on the order I have in mind which is:

- Update the base/broker.go to include a shared init that will bootstrap the instance.
 - This could get a bit messy in that each instance type has attributes specific to that service--e.g., adapter for RDS.
 - An approach here could be to write those attributes as a json string in a "properties" column. I am open to other approaches. Just trying to allow for maximum flexibility so it can scale to allow any AWS service.
 - Another approach could be to have things like Adapter be it's own table and have a join on the instances table in the column of "associations" or something along those lines.
 - Again, definitely would love your feedback on how you'd like this done.
- Augment the instances table to include a column for type, properties, etc based on the implementation above.
- Implement a shared IAM package for use in the CRUD operations of each broker.
 - This is one of those massive overlap cases.
 - Probably leverage this as much as possible: https://github.com/cf-platform-eng/sqs-broker/tree/master/awsiam
- Refactor the existing RDS broker to use the new IAM package and make sure everything works correctly.
- Add in SQS functionality

Does that sound about right?

Thanks!

**Comment by jcscottiii** [comment]
@scottt2
That plan sounds perfect to me!
My only note: Let's avoid putting all the extra properties in the a properties field.

So happy you brought up all these great points @scottt2

**Comment by scottt2** [comment]
Hey @jcscottiii ! Awesome, man. It's been great bouncing ideas off you. Thanks for the dialog.

For the properties, do we want each to have their own table then?

For example, upon creation, an RDS instance would be written to the `instances` table and in that table would be a column called `properties` that contains an id for an entry in an `rds_properties` table--i.e., a hasone association of instances to properties. We'd have to do a bit of logic based on instance type to determine the correct properties table. I don't think it would be too complex, though. I think I'd need to start writing it to fully comprehend.

I *think* that would be most scaleable. Thoughts?

**Comment by jcscottiii** [comment]
@scottt2 That sounds good
The Instance entry will be generic for all types of instances.
You might also need to add another column in `instances` named `InstanceType`. That will tell you which table to query the id. (e.g rds_properties)

**Comment by scottt2** [comment]
Good call, @jcscottiii . I think that would be a good idea. I was originally going to infer type based off `service_id` that's stored in `instances`. Storing it explicitly would reduce the logic for finding the table.

**Comment by jcscottiii** [comment]
@scottt2 another idea (you might have already done this) to build on what i said. That type field I just mentioned should be an enum of some sorts e.g

```go
type InstanceType uint16
const (
        RDS InstanceType = iota
        SQS
        // Other ones
)
type Instance struct {
   // other fields
   Type InstanceType `some struct tags`
}
```
It will prevent you from having to do string comparison and give you some type of type safety.

but do whatever best fits.

**Comment by scottt2** [comment]
@jcscottiii Nice! Haven't implemented yet. Trying to wrap other projects. Will do. Thanks!

**Comment by scottt2** [comment]
Morning/Afternoon @jcscottiii @mogul @dlapiduz !

I was banging my head against the wall on merging the SQS Broker into this repo and had an idea--what if we started with a V2 broker based of the pattern you linked above, @jcscottiii and rolled another V2 broker into it making a new AWS Broker?

Here is the experiment so far: https://github.com/scottconstructioninc/sqs-broker/pull/1

I have some more details in that PR. Having the brokers based off the same pattern makes this *vastly* easier to reason about. We also benefit in having simplified router logic (yet to be built into that PR). Everything is easy when they are the same basic structure.

As noted in the PR, doing this would mean this repo would need a major version bump and messaging would have to be provided to users upgrading that there may potentially be breaking changes. I haven't gotten it all wired up yet to test for parity with the existing, so I'll know more once that happens.

Anyhow, I'd really love to hear your thoughts on this approach.

Thanks!

**Comment by jcscottiii** [comment]
@scottt2 
What things were breaking?
I'm looking through the code here. I think the biggest thing I saw was trying to do configs. (I saw you implemented our broker interface but when it came to getting the settings things get a little hairy. I think it would be easier to maintain if we use configs in YAML format too (we moved away from using JSON but left all the tags in there, I have a branch here that I do some work on from time to time. don't worry about it but it has an example of the YAML config https://github.com/18F/aws-broker/blob/tests/catalog/catalog_test.go). I'll add some comments where I'm talking about.

Pulling in the other RDS broker is a big change :-/ and would like to stick to the original plan of augmenting this current broker.

File: catalog/catalog.go [hunk 1]
```diff
@@ -57,24 +214,27 @@ var (
 )
 
 // RDSService describes the RDS Service. It contains the basic Service details as well as a list of RDS Plans
```
Comment (jcscottiii) [inline catalog/catalog.go]: You could create a SQSService also here.

Comment (jcscottiii) [inline catalog/catalog.go]: Youll need a `FetchPlan` method too.

File: manager.go [hunk 2]
```diff
@@ -1,21 +1,28 @@
...
+	"github.com/18F/aws-broker/services/sqs"
 	"github.com/jinzhu/gorm"
-	"net/http"
 )
 
 func findBroker(serviceID string, c *catalog.Catalog, brokerDb *gorm.DB, settings *config.Settings) (base.Broker, response.Response) {
+	fmt.Println(serviceID)
 	switch serviceID {
 	// RDS Service
-	case c.RdsService.ID:
```
Comment (jcscottiii) [inline manager.go]: you will have a separate case for the `SQSService`

**Comment by scottt2** [comment]
@jcscottiii  Yeah, definitely agree it is a massive change. Though in what I'm finding in my local, it's also a pretty big change. I haven't push up any of the latest yet because the place is a mess while I'm figuring things out.

The biggest issues for me were:
* Adding in shared IAM policies. The existing RDS Broker doesn't have this. It uses the AWS session package, so augmenting the existing RDS Broker to use the IAM so we can have something like this: https://github.com/cf-platform-eng/rds-broker/blob/master/iam_policy.json is a bit of work.
 * If we forego using that for RDS, we'd still need it for SQS per the spec. The added benefit here being I could conceivably use much of the existing SQS Broker code. Once in place, it would make sense to have RDS use it as well.
* Shared catalog is awkward.
 * In the other branch, each service contains it's own catalog and all the structs that map to it. Having one central yaml/json is awkward. This becomes a problem with things that are service specific being mapped into a generic placeholders that we then later parse out into structs. This dynamic assignment is difficult to do cleanly.
* SQS Broker is using the published V2 broker api pattern whereas the existing AWS Broker is using an older version.
 * This makes having a shared router difficult as things don't align very well. For example:
  * https://github.com/18F/aws-broker/blob/master/base/broker.go
  * compared with
  * https://github.com/pivotal-cf/brokerapi/blob/master/service_broker.go
 * What I am finding is I need to significantly refactor the existing broker base to better match the SQS Broker or I need to write a wholly new SQS Broker that aligns with the previous api pattern.
 * In my estimation, it is less work to do the former than the latter, but both seem less optimal than using the community provided V2 brokers for each.

That's off the top of my head. If the other approach is a non-starter, I'd likely pull fresh from develop here and cherrypick in the things that were good from what I've done so far and continue refactoring the existing broker to better match the pattern of the V2 broker.

RE: the .json/.yml--I went with json because I saw all the tags in there already and due to Go's decent handling of JSON in the stdlib. I can switch to yml, no problem--just curious on the preference as I'm ignorant to benefits.

**Comment by jcscottiii** [comment]
@scottt2 Apologies for the late reply.

Let's go for the refactor where needed. If importing / re-creating some of the input structs / functions will help, let's do it. While the existing SQS broker (which comes with a lot of great bells and whistles) is there, not everything needs to come directly from there. The success of our RDS broker has been due to the simplicity (but as you can see it's outdated now lol)

For the shared catalog, things won't be generic. It's impossible. You've seen that too and mentioned that lol.. With the YAML you can have this to handle the specifics and read by the root keys of the YAML.
```
rds:
  - # details specific for rds
sqs:
  - # details specific for sqs
```

For the IAM stuff, that will probably have to be done from scratch. I know that's not the best answer but I don't know how else to go about that.

**Comment by scottt2** [comment]
No worries, @jcscottiii ! You've been awesome to work with.

I totally get it. We want to keep as much legacy as possible. It is definitely the safer route. I'll focus on the refactor where needed, then. If you ever want the other solution, it's in that repo.

RE: catalog--I was hoping for something better than just relying on omit when empty and having every service in the catalog carry with it the structs for each instance type. Am I thinking about that wrong? The generic Service would need to account for any potential service specific settings, so it would need to have a data structure to handle it if it's present.

Bummer on the IAM stuff...That is the thing that's really looming on my remaining time before deadline. I'll see what if anything I can pull in and go from there.

Regardless of where I net out at 5pm EST on Friday, I sure appreciate all your help on this, @jcscottiii . My lesson learned on these short turn around jobs is to engage folks like yourself as early as possible so I can account for the true scope of the effort!

Thanks again, man!

**Comment by jcscottiii** [comment]
@scottt2, 
i know what you mean about the catalog. i think that will be how it will need to be for now. with that, you wouldn't need to have a generic service either.
```go
// Catalog struct holds a collections of services
type Catalog struct {
	// Instances of Services
	RdsService RDSService `yaml:"rds" json:"-"`
	SqsService SQSService `yaml:"sqs" json:"-"` // Has all your service specific stuff

	// All helper structs to be unexported
	secrets   Secrets   `yaml:"-" json:"-"`
	resources Resources `yaml:"-" json:"-"`
}
```
let me know if i missed that point though (getting late)!

it's new for our team too. we want this to be great experience for everyone :) I'm really happy that we are having this back and forth conversation and not just getting something randomly at the end. it really has been a pleasure!

File: common/dbconfig.go [hunk 3]
```diff
@@ -43,16 +44,9 @@ type DBConfig struct {
...
-		log.Printf("Attempting to login as %s with password length %d and url %s to db name %s\n",
-			dbConfig.Username,
-			len(dbConfig.Password),
-			dbConfig.URL,
-			dbConfig.DbName)
-	*/
-	switch dbConfig.DbType {
+	//var DB gorm.DB
+	//var err error
+	/*switch dbConfig.DbType {
```
Comment (scottt2) [inline common/dbconfig.go]: **Note:** This is just a temp comment block. My local env is all sorts of messed up right now, so I just skip and go right to sqlite3.

**Comment by mtorres253** [comment]
Hi all, just wondering where we are on delivery deadline? Any concerns?

**Comment by scottt2** [comment]
Hey @mtorres253 !

I am not going to make the deadline to get the award.

I did not properly spec out the amount of work necessary for completing this task in the time I had--which is embarrassing, but hopefully what I did get done helps the next person.

Noting here for future self or any other 18F bidders--engage folks like @jcscottiii (you were super helpful, man!) before you sign off on your bid. I know it's difficult sinking time into a bid before you have won, but a cursory reading of the spec is definitely not enough. I could have saved myself alot more unpaid hours by realizing the true scope and passing because I simply didn't have the time to do it the way it was specced. 

Anyhow, I went through and cleaned up the code so the next in line can have a decent starting point. Here is where we are at:

**TL;DR**--We need the guts of the SQS and IAM wired up.

✅ ~~Scenario: Request is made to view SQS options~~

~~Given there is one service plan for the SQS service~~
~~When a request is being made to view the broker’s catalog~~
~~Then the response should include the SQS service~~
~~And the response should include the name of the SQS service plan~~
~~And the response should include a description of the service plan~~
~~And the response should include the pricing associated with each SQS service plan~~
~~And the response should be in [JSON] format.~~
~~Scenario: New service instance with SQS is created~~

:white_medium_square:~~Given there is one SQS plan~~
~~When a new service instance is provisioned with a specified SQS opti~~
Then a new service instance is created for that plan
And an SQS queue is created and associated with that instance
Scenario: An app gets valid credentials when bound to the SQS service instance

* **Notes**
 * Instance is saved to the db, but no associated SQS queue is created.
 * Will need to finish building the SQS and IAM logic.

:white_medium_square:Given there is an SQS service instance
When an app id is bound to the instance
Then new IAM user credentials are returned for that app
And that IAM user has access to the queue associated with the instance
Scenario: An app id’s credentials are invalidated when unbound to the SQS service instance

* **Notes**
 * See above. Blocked by function SQS and IAM logic.

:white_medium_square:Given there is an SQS service instance
And an app id that is bound to the instance
When the app id is unbound from the instance
Then the app id’s IAM user credentials no longer have access to the queue associated with the instance
Scenario: Removing service instance with results in deletion of associated SQS queue and IAMs

* **Notes**
 * See above. Blocked by function SQS and IAM logic.

:white_medium_square:Given there is an SQS service instance
And there is an app id bound to the service instance
When the service instance is removed
Then the associated SQS queue is deleted
And any IAM credentials issued for that queue are revoked

* **Notes**
 * See above. Blocked by function SQS and IAM logic.

**Comment by mtorres253** [comment]
How much more time would you estimate it would take to complete the work? How confident are you in the estimate? @scottt2
